### ISSUE TITLE
I have question about MultipleNegativesSymmetricRankingLoss  (.foward)  

### ISSUE THREAD

**Issue by sogm1** [UKPLab/sentence-transformers#2508]
Hello 

I recently use study sentenceTransformer loss , i have question about loss .forward in ( MultipleNegativesRankingLoss ,MultipleNegativesSymmetricRankingLoss)

Here is MultipleNegativesSymmetricRankingLoss 
```

import torch
from torch import nn, Tensor
from typing import Iterable, Dict
import sys
sys.path.append("/content/drive/MyDrive/논문/sentence_transformers")
from sentence_transformers import SentenceTransformer
from sentence_transformers import util

class MultipleNegativesSymmetricRankingLoss(nn.Module):
    def __init__(self, model: SentenceTransformer, scale: float = 20.0, similarity_fct=util.cos_sim):
        """
        This loss is an adaptation of MultipleNegativesRankingLoss. MultipleNegativesRankingLoss computes the following loss:
        For a given anchor and a list of candidates, find the positive candidate.

        In MultipleNegativesSymmetricRankingLoss, we add another loss term: Given the positive and a list of all anchors,
        find the correct (matching) anchor.

        For the example of question-answering: You have (question, answer)-pairs. MultipleNegativesRankingLoss just computes
        the loss to find the answer for a given question. MultipleNegativesSymmetricRankingLoss additionally computes the
        loss to find the question for a given answer.

        Note: If you pass triplets, the negative entry will be ignored. A anchor is just searched for the positive.

        :param model: SentenceTransformer model
        :param scale: Output of similarity function is multiplied by scale value
        :param similarity_fct: similarity function between sentence embeddings. By default, cos_sim. Can also be set to dot product (and then set scale to 1)

        Requirements:
            1. (anchor, positive) pairs

        Relations:
            - Like :class:`MultipleNegativesRankingLoss`, but with an additional loss term.

        Inputs:
            +---------------------------------------+--------+
            | Texts                                 | Labels |
            +=======================================+========+
            | (anchor, positive) pairs              | none   |
            +---------------------------------------+--------+

        Example:
            ::

                from sentence_transformers import SentenceTransformer, losses, InputExample
                from torch.utils.data import DataLoader

                model = SentenceTransformer('distilbert-base-uncased')
                train_examples = [
                    InputExample(texts=['Anchor 1', 'Positive 1']),
                    InputExample(texts=['Anchor 2', 'Positive 2']),
                ]
                train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=32)
                train_loss = losses.MultipleNegativesSymmetricRankingLoss(model=model)
                model.fit(
                    [(train_dataloader, train_loss)],
                    epochs=10,
                )
        """
        super(MultipleNegativesSymmetricRankingLoss, self).__init__()
        self.model = model
        self.scale = scale
        self.similarity_fct = similarity_fct
        self.cross_entropy_loss = nn.CrossEntropyLoss()

    def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor):
        reps = [self.model(sentence_feature)["sentence_embedding"] for sentence_feature in sentence_features]
        anchor = reps[0]
        candidates = torch.cat(reps[1:])

        scores = self.similarity_fct(anchor, candidates) * self.scale
        labels = torch.tensor(
            range(len(scores)), dtype=torch.long, device=scores.device
        )  # Example a[i] should match with b[i]

        anchor_positive_scores = scores[:, 0 : len(reps[1])]
        forward_loss = self.cross_entropy_loss(scores, labels)
        backward_loss = self.cross_entropy_loss(anchor_positive_scores.transpose(0, 1), labels)
        return (forward_loss + backward_loss) / 2

    def get_config_dict(self):
        return {"scale": self.scale, "similarity_fct": self.similarity_fct.__name__}
```

SentenceTransformer use  InputExample.
Maybe  i guess if batch_size = 32 
 , it would be  
```
  train_examples = [
                    InputExample(texts=['Anchor 1', 'Positive 1']),
                    InputExample(texts=['Anchor 2', 'Positive 2']),
                    InputExample(texts=['Anchor 3', 'Positive 3']),
                     ... 
                    InputExample(texts=['Anchor 32', 'Positive 32'])
                ]
```
and train_dataloder transfer this for model in batch.  

**Q1. Does reps[0] means [Anchor 1,Anchor 2 ...  Anchor 32]   and reps[1:] means [Positive1,Positive2... Positive32] ?** 
- I don't know if what I'm thinking is the right answer

```
 def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor):
        reps = [self.model(sentence_feature)["sentence_embedding"] for sentence_feature in sentence_features]
        anchor = reps[0]
        candidates = torch.cat(reps[1:])

```

**Q2. What is shape of  `anchor_positive_scores = scores[:, 0 : len(reps[1])]`**  
- If batch_size =32  , i guess  tensor shape of scores  is 32 x 32,  that means  len(reps[1]) = 32? 
-  so , shape of anchor_positive_scores  is 32 x 32 , is it same as scores?  i am confused 😂

**Q3. Could i get  example code for above that? i hope print(shape of tensor) is very helpful for understanding.**

Thank you  
Best regards

**Comment by tomaarsen** [comment]
> I finally understood after sentenceTransformer class. So far, I got it wrong.
> 
> reps[0] shape = (batch, dim) # anchor reps[1:] shape = (batch, dim) # positive ,if with negative, batch= batch*2
> 
> score shape = [batch, batch]
> 
> During Traning , the diag number of score matrix get higher caused by cross-entropy make loss

That is right! Apologies that I didn't get around to answering your question earlier. I have been busy these last few days.

- Tom Aarsen
